aperturedb/SPARQL.py

```diff
@@ -0,0 +1,607 @@
+            if "list" not in command["results"]:
+                command["results"]["list"] = []
+            prop = self._parse_uri_with_ns("p", p)
+            command["results"]["list"].append(prop)
+            bindings.append((o, command, prop))
+
+        def add_is_connected_to(s, p, o):
+            """Adds connection constraint between two commands"""
+            s_command = commands[s]
+            o_command = commands[o]
+            if s_command["_ref"] < o_command["_ref"]:
+                command = o_command
+                ict = dict(
+                    ref=s_command["_ref"],
+                    direction="out",
+                )
+                connection_bindings.append((o_command, p, s_command, False))
+            else:
+                command = s_command
+                ict = dict(
+                    ref=o_command["_ref"],
+                    direction="in",
+                )
+                connection_bindings.append((s_command, p, o_command, True))
+
+            if p is not None:
+                connection_type = self._parse_uri_with_ns("c", p)
+                ict["connection_class"] = connection_type
+            else:
+                del ict["direction"]  # Any connection, either in or out
+
+            if "is_connected_to" not in command:
+                command["is_connected_to"] = ict
+                command["group_by_source"] = True
+            elif list(command["is_connected_to"].keys())[0] == "all":
+                list(command["is_connected_to"].values())[0].append(ict)
+            else:
+                command["is_connected_to"] = {
+                    "all": [command["is_connected_to"], ict]}
+
+        def process_bindings(output, ctx: "QueryContext" = ctx, i=0, ids=set()):
+            """Processes AperatureDB response and yields bindings
+
+            Operates recursively, processing each command in the query to generate bindings
+            for each variable, multiplying the number of bindings at each step.
+            Uses `solutions` to avoid yielding duplicate solutions.
+            """
+            from rdflib.term import Variable, BNode, Literal, URIRef
+            if i == len(output):
+                solution = ctx.solution()
+                # print("Yielding solution", solution)
+                if solution in solutions:
+                    return
+                solutions.add(solution)
+                yield solution
+                return
+
+            command_body = list(query[i].values())[0]
+            result_body = list(output[i].values())[0]
+            entities = result_body.get("entities", [])
+
+            bindings2 = [(v, prop) for v, cb, prop in bindings
+                         if cb == command_body and not isinstance(v, BNode)]
+            assert all(isinstance(v, Variable)
+                       for v, _ in bindings2), bindings2
+            assert all(ctx[v] is None for v, _ in bindings2), bindings2
+            # print(
```

Comment:
We could logger debugs this .

Reply:
That specific print isn't very useful, but I'll add some logging.